v1rc (#10)
fallen-icarus/cardano-swaps: README.md

Cardano-Swaps is an order-book based DEX that takes inspiration from Axo's programmable swaps
[design](https://www.axo.trade/whitepaper.pdf), adds delegation control as a foundational feature,
naturally incentivizes its own liquidity, and, through the use of Beacon Tokens, removes the need for
specialized indexers. In short, it solves all of the pressing issues that made order-book DEXs
inferior to LP based DEXs. 

### High-Level Overview

The protocol is comprised of two types of composable swaps: a one-way swap and a two-way swap.
One-way swaps are effectively just limit orders (ie, they can only be executed at the specified
price or better). Most users will likely use the one-way swap.

Two-way swaps are a mechanism for profitably providing liquidity to the protocol without having to
resort to yield farming. These two-way swaps can go in either direction (eg, a two-way ADA <-> ERGO
swap can go ADA -> ERGO or ERGO -> ADA). By allowing the liquidity provider to specify separate
prices for each direction, they can profit from either direction of the swap. For example, if the
market rate for DJED <-> USDC is 1:1, Alice can set DJED -> USDC to 1.01 USDC per DJED and USDC ->

> DJED to 1.01 DJED per USDC. This means Alice makes a 1% return no matter what direction is used.

Author: @rphair should these be:

> Alice can set DJED -> USDC to 1.01 usdc per djed...

Reviewer: They're being quoted as trading symbols & in exchange rates, so the convention would be to uppercase them.

sea of available swaps.

Since stake pool operators are required to always be connected to the network, they are uniquely
positioned to serve as arbitragers. It can provide another potential source of profit for all stake
pools operators, including the small pool operator who rarely makes blocks.

To fully appreciate the implications of the two-way swaps, it helps to consider a few examples:

##### The Contrived Arbitrage Example

``` Txt
Alice has 10 ada in her swap address and is willing to swap them for 0.5 AGIX/ADA.
Bob has 10 agix in his swap address and is willing to swap them for 1 ADA/AGIX.
```

In this example, Charlie can profitably arbitrage and fulfill both of these swaps like this:

``` Txt
Charlie looks up all swap addresses willing to swap AGIX/ADA. Charlie finds Alice's address.
Charlie looks up all swap addresses willing to swap ADA/AGIX. Charlie finds Bob's address.
Charlie gives Bob 10 ada and receives 10 agix.
Charlie gives Alice 5 agix and receives 10 ada.
Charlie now has his original 10 ada plus an additional 5 agix.
This all occurs in one transaction where Charlie pays the transaction fee.
```

> On net, Charlie pays the transaction fees and receives 5 AGIX in return, while both Alice's and

Reviewer: YMMV with other cryptocurrencies.  We'll probably encounter some other ones which insist that the cryptocurrency are units are capitalised, to match common usage based on obsession with crypto markets & trading.

We can't please everyone let alone match this divergence so I think the community would leave it up to you.  The people employing correct standards are often in the minority & I wouldn't want to place your material at a disadvantage for it: even if the matter is clear for `ada` vs. `ADA`.

For instance I followed the EU guidelines to call the new currency `euro` (not capitalised, always singular) since before the notes were even printed: then watched for 2 decades while everyone called the notes & values "euros" and/or Euro even in government communications. 🤪

> On net, Charlie pays the transaction fees and receives 5 AGIX in return, while both Alice's and

Reviewer: That sounds reasonable enough for me.  It's the Cardano Foundation that makes the repeated calls about under-capitalising the currency unit.  If & when it comes up in CIPs, other standards discussions & the community I'll do what I can do help justify the use of ADA if you choose... for the general reason that the platform itself is centred on trading.

Sarah gives Bob 5 agix and receives 10 dust.
Sarah gives Alice 10 dust and receives 10 ada.
Sarah now has her original 10 ada plus an additional 5 agix.
This all occurs in one transaction where Sarah pays the transaction fee.
```

On net, Sarah pays the transaction fee and receives 5 agix in return, while four swaps are
fulfilled. As shown in this example, Sarah fulfills *both* the AGIX/DUST swap and the HOSKY/AGIX
swap by "passing through" those pairs on her way back to ada. As long as the entry and exit pairs
(in this case ADA/HOSKY and DUST/ADA) have enough liquidity, arbitragers can spread that liquidity
into less liquid swap pairs. And since two-way swaps naturally incentivize providing liquidity for
major trading pairs, this requirement for entry and exit liquidity is naturally incentivized to be
satisfied.

Since the current design is capable of handling 11-14 swaps in a single transaction, there is ample
flexibility for finding arbitrage opportunities. For example, maybe the 3rd swap is at a slight loss
but the 5th swap is such a good arbitrage that it more than makes up for it resulting in a net
profit for the arbitrage.

##### The Stablecoin <-> Stablecoin Two-Way Swap

> Currently, DApps treat stablecoins individually (ie, you may be able to borrow a loan in DJED but

Reviewer: https://github.com/fallen-icarus/cardano-swaps/pull/10#discussion_r1392731880

> not USDC). This means, even though all stablecoins are practically the US dollar, the liquidity is

Author: USDC -> usdc

Reviewer: https://github.com/fallen-icarus/cardano-swaps/pull/10#discussion_r1392731880

fractured across all stablecoins. Two-way swaps change this.

Imagine if Charlie has a swap for DJED <-> USDC where converting assets in either direction requires
paying a 1% premium. The very existence of this swap, combined with the fact that Cardano-Swaps is
freely composable with all DApps, means users can freely convert between stablecoins as needed. The
liquidity would no longer be fractured. Alice could convert her DJED to USDC to buy an options

> contract that is asking for USDC. She just has to pay Charlie the 1% fee for the conversion.

Author: Same here.

Since two-way swaps can go in either direction, these swaps do not need to be reset after each swap.
For example, after Alice deposits DJED into the swap to claim USDC, Bob can now use the swap to

> convert his USDC to DJED by paying a 1% fee. Either way, Charlie makes 1% per swap and *is paid

Author: Here too.

Reviewer: https://github.com/fallen-icarus/cardano-swaps/pull/10#discussion_r1392731880

directly in the stablecoin being deposited*. And since the stablecoins are supposed to be 1:1 with
the US dollar, the only risk is a peg-break. Anyone can be a liquidity provider and profit like
Charlie does in this example.

> Being able to treat all stablecoins as a single stablecoin will be a game changer for DeFi.

Author: I am not sure if I should add something after this line. AFAICT this ability to unify stablecoin liquidity is unique to Cardano-Swaps *because* it is p2p and natively composable. I don't think either Axo or GeniusYield can do this since their batchers would need to execute the transaction for you. It is the same as the "batchers can't trustlessly compose arbitrary DApps" argument. They may support some compositions but not others which would mean some stablecoins conversions are still not possible. This asymmetry of support for conversion may be worse than not supporting conversions at all; it effectively means Axo and GeniusYield can choose winners and loses among stablecoins.

##### The Volatile Two-Way Swap

While the stablecoin <-> stablecoin two-way swaps are the least risky, there is also an incentive to
provide liquidity for volatile pairs. Consider two assets like ada and dust. Imagine if Alice thinks
ada is going to appreciate against dust; therefore, she would rather have ada. Her prices can
reflect this: converting ADA -> DUST could require a 15% premium while converting DUST -> ADA could
require only a 5% premium. Would anyone pay the 15% premium?

Yes, arbitragers would. Imagine if an arbitrager found a really good opportunity but they needed the
ada. As long as the overall profit from the arbitrage can cover the 15% premium (and then some), the
arbitrager may be more than willing to pay this high premium. Considering that the protocol supports
chaining up to 14 swaps and illiquidity implies great arbitrage opportunities, it is actually likely
that arbitragers will be able to find opportunities that justify paying this high premium. For

> example, let's say a single arbitrage yields a 5% profit and, since 14 swaps can fit in a single

Author: I think I should remove this abstract example. The following concrete example seems easier to understand and seems to make the case better. I don't think I need both examples.

possibly infinite set of addresses for users.

### Democratic Upgradability

Upgrades to Cardano-Swaps can propagate through the ecosystem of users in a similarly democratic
fashion as SPOs upgrading their pools to a new version of `cardano-node`. Since users can close
their swaps at any time, whenever there is a potential upgrade, users can choose to close their
current swaps and recreate them with the new contracts. There is never a bifurcation of liquidity
due to the composable nature of all swaps, regardless of the protocol's version.

### Frontend Agnosticism

Thanks to the query-ability of beacon tokens, it is trivial for any frontend to integrate with
Cardano-Swaps. For example, any wallet can integrate `cardano-swaps` by adding support for querying
the beacon tokens. They can also add their own user friendly way to create and use swaps. The only
requirement is that all frontends/users agree to use the same beacon token standard. There is no
need for dedicated (censorable) frontends.

### Naturally Resistant to Denial-of-Service Attacks

Recall that each swap requires a deposit of at least 2 ada. While this behavior is due to the

> current protocol parameters, it is actually a good thing since it helps prevents denial-of-service

Author: "helps prevents" -> "helps prevent"

attacks. For example, since the protocol does not check that the assets in the datum are real
assets, it is possible to create swaps for fake assets. *This does not impact normal users who are
querying swaps for a known trading pair.* The user would have to deliberately query a trading pair
that contains a fake asset in order to see these fake UTxOs.

These fake asset UTxOs only really impact offer and ask queries which are mainly for arbitragers who
will likely have powerful hardware. Since these fake UTxOs still require 2 ada, creating millions of
them (enough to disrupt the arbitragers) would require a deposit of millions of ada throughout the
life of the denial-of-service attack. The only way to reclaim the ada is to stop the attack.
Furthermore, checking if an asset is real is trivial (look for mint history). Once an arbitrager
realizes an asset is fake, they can ignore it. So an actual denial-of-service attack would require
constantly changing the millions of fake swaps to new fake swaps; this has a large cost in terms of
transaction fees which makes this impractical to do at scale.

> ## FAQ

Reviewer 2: I'd add that in the endgame, _most_ users will not be "competing" for swaps - market makers / businesses will be. Most users will create orders via swap UTxOs that are then swept by larger players (MMs/arbitragers) - in this scenario there is zero concurrency problem.

> ## FAQ

Reviewer 3: I respectfully would recommend that you simplify the explanation somehow because it seems like the description is making things sound worse than they really are.  The way I see things is that market makers are going to be searching for all opportunities to profit and therefore they will combine all the overlapping djed swaps that can be executed profitably.  I think the market makers will end up competing for essentially the same "meshwork" of profitable swaps they want to combine.  This is a slight irritation for the market makers who will see some transactions fail, particularly highly profitable ones that will get competed over.  But, it is not a problem for ordinary users who will quickly get their swaps executed if "in the money".

> ## FAQ

Author: How about this wording?

> ### If there is one two-way swap UTxO with 1000 djed and there are 10 users trying to use 100 djed each, isn't this still a concurrency issue despite having enough liquidity?                                     
>                                                                                                                                                                                                             
> Yes, this is a concurrency bottleneck but no, this is not an issue. This is the perfect niche for businesses to build on top of Cardano-Swaps. Historically, whenever there is a risk that users don't want to carry, a business can form to carry that risk instead, for a price. Insurance is one such example. The same scenario applies here.                                                                                                                                                                           
>                                                                                                                                                                                                                   
> Imagine a slightly simpler example: Alice has a 100 djed swap available while Mike and Charlie both want 50 djed from it. No liquidity bottleneck here, just a concurrency bottleneck. Sarah can see  this scenario and decide to front the 50 djed to both Mike and Charlie (possibly for a fee) with the intention of settling up with Alice's swap later, ideally for a profit. At a high-level, this is how market making works. Mike and Charlie will likely prefer going through Sarah, even if Sarah charged a fee, simply because it means they don't have the concurrency risk. *The option to directly interact with Alice's swap will always be available to Mike and Charlie.* Because this option is always available, Sarah is incentivized to give Mike and Charlie a fair deal. Tiered fees will make it even easier for Mike or Charlie to use Alice's swap directly, if needed.                                                                                                                                          
>                                                                                                                                                                                                                   
> Thanks to Cardano-Swaps being fully p2p and permissionless, anyone can build a business on it like Sarah. And since the day-to-day users will likely not want the concurrency risk, there is every incentive for people like Sarah to step in. In the endgame, *most* users will not be competing for swaps - market makers and other businesses will be. Most sellers will create orders via swap UTxOs that are then swept by larger players (MMs/arbitragers) while most buyers will go through businesses that carry the concurrency risk in their stead. In this scenario, there is zero concurrency problem for the end user.                                                                                                                                                                                                  
>                                                                                                                                                                                                                   
> When it comes to creating a decentralized economy, not every feature needs to be directly enabled by the DApp. Some features can, and should, be left to the market to fill.

> ## FAQ

Reviewer: Both work equally well for me but I think the latter (shorter, more familiar) edit will be more approachable by general readers... even docs for technical audiences in crypto & other Internet apps still have the Alice & Bob stories 😄

> ## FAQ

Reviewer 3: In my head, I see things playing out like this:

User Alice has a 100 djed swap available.  User Mike and user Charlie both want 50 djed from it, so they submit their swaps.  A market maker sees these three swaps on-chain and notices that the exchange price that both Mike and Charlie have set overlaps with the price that Alice wants.  Consequently the market maker submits a transaction that uses Alice's swap combined with both Mike's and Charlie's so that all three swaps are executed in one transaction and the market maker profits from the amount of price overlap minus the transaction fees.

Thus there isn't a concurrency issue because market makers will look for such trades and try to bundle as many as possible that overlap into one transaction whenever a profit can be made.  The only concurrency issue results from the market makers competing to be first to execute the entire linked bundle of profitable (overlapping price) trades.

Furthermore, since anyone can be a searcher or market maker, anyone can search for, bundle up, and execute, profitable trades.  The more that are profitable which can be combined into one transaction the lower the proportional fees for the searcher.

> ## FAQ

Author: > User Alice has a 100 djed swap available. User Mike and user Charlie both want 50 djed from it, so they submit their swaps. A market maker sees these three swaps on-chain and notices that the exchange price that both Mike and Charlie have set overlaps with the price that Alice wants.

This only works if Charlie and Mike both deliberately incentivize someone to execute their swaps. To put it concretely, they can't use the exact same price as Alice. They need to deliberately create an arbitrage opportunity. To use the example, Mike and Charlie can only ask for 49 djed each so that 1 djed can be claimed by the arbitrager for each swap.

I think this is valid, though. I hadn't considered deliberately creating an arbitrage opportunity. I always thought of arbitrage opportunities as an accident. For volatile pairs, you *might* be able to use the same price but if the main concern is immediate conversion, the better the incentive for arbitragers the faster it will be converted. 

I think I like this approach more even though both are valid. With this one, Charlie and Mike set the fee they are willing to pay for immediate conversion. Meanwhile, in the previous example, Sarah sets the fee.

Nice observation!

> ## FAQ

Author: @TerminadaPool I changed it to use your method. How's this:

> ### If there is one two-way swap UTxO with 1000 djed and there are 10 users trying to use 100 djed each, isn't this still a concurrency issue despite having enough liquidity?                                     
>                                                                                                                                                                                                              
> Yes, this is a concurrency bottleneck but no, this is not an issue. This is the perfect niche for businesses to build on top of Cardano-Swaps. Historically, whenever there is a risk that users don't want to carry, a business can form to carry that risk instead, for a price. Insurance is one such example. The same scenario applies here.                                                                                                                                                                           
>                                                                                                                                                                                                              
> Imagine a slightly simpler example: Alice has a 100 djed swap available while Mike and Charlie both want 50 djed from it. No liquidity bottleneck here, just a concurrency bottleneck. If Mike and Charlie do not want to deal with the concurrency issues, they can pay arbitragers to execute the swaps for them. Here is how it would work:                                                                                                                                                                         
>                                                                                                                                                                                                                 
> - Both Mike and Charlie create one-way swaps with prices so that, if executed against Alice's swap, Mike and Charlie would each get 49 DJED and whoever executes their swaps gets to keep the other 1 DJED per swap. These swaps are effectively saying: "I'll pay you 1 DJED to execute this swap for me."                                                                                                                                                                                                               
> - Sarah, an arbitrager, can see all three swaps and executes them, keeping the 2 DJED arbitrage (1 DJED from each swap).                                                                                                                                                                                              
>                                                                                                                                                                                                                
> In other words, Mike and Charlie *deliberately create an arbitrage opportunity* so that arbitragers will want to execute their swaps. If Charlie really needs his swap executed ASAP, he can set his price so that he gets 48 DJED and the arbitrager gets 2 DJED for executing his swap. Basically, the more attractive the buyer makes the arbitrage opportunity, the more arbitragers will compete for it and the faster their swap will be converted. In this scenario, neither Mike nor Charlie experience a concurrency bottleneck. The arbitragers competing for their swaps are the ones experiencing the concurrency bottleneck, and they are paid accordingly for taking on this risk.                                                                                                                                     
>                                                                                                                                                                                                      
> In the endgame, *most* users will not be competing for swaps - arbitragers and other businesses will be. Most users will create orders via swap UTxOs that are then swept by these businesses. In this scenario, there is zero concurrency problem for the end user.                                                                                                                                                      
>                                                                                                                                                                                                          
> When it comes to creating a decentralized economy, not every feature needs to be directly enabled by the DApp. Some features can, and should, be left to the market to fill.

> ## FAQ

Author: This is now the second question in the FAQ. I added a short paragraph noting the similarities with setting a limit order below the market price in TradFi. I personally did not notice this connection until @TerminadaPool brought it up so I don't want to assume others will immediately see the connection.

> ## FAQ

Reviewer 3: I still think you are underselling Cardano-Swaps.  I love your project, so this is why I am trying to figure out how you can word this FAQ without underselling it.

How does this sound?:

Imagine a slightly simpler example: Alice has a 100 DJED -> 100 ADA swap available.  Mike and Charlie both want to swap 50 ADA for DJED.  If they post their swaps on-chain as follows 50 ADA -> 50 DJED then there is no profit motive for anyone to combine and execute these trades after including the small transaction fee.

Mike and Charlie can easily each create a new swap which consumes Alice's swap where they pay the transaction fee.  However, if they don't want to do the combining themselves they can easily create a swap with enough overlap in price so that an arbitrager will be incentivised to bundle up and execute their trades.  For example, Mike and Charlie could both create 49.8 ADA -> 50 DJED swaps.  This would enable an arbitrager to notice that there is a 0.2 ADA profit opportunity (less transaction fee).  The arbitrager can now combine all 3 swaps for a slight profit.  He only needs 98.6 ADA to get 100 DJED from completing both Mike's and Charlie's swaps and can use the 100 DJED he will receive to execute Alice's swap to receive 100 ADA.  This will net the arbitrager 0.4 ADA less approx 0.2 ADA transaction fee for a net profit of 0.2 ADA.  One thing to notice about this is that the more swaps the arbitrager can combine, the more profitable it can be since he will always only pay one transaction fee and this fee won't increase very much with each extra swap added.  Arbitragers will seek out and complete these swaps because even when the profit is small because there is no risk to them and it is free money.

Many users of Cardano-Swaps will simply list their swaps on-chain and wait for their set price to be reached, just like Alice in the example above who created a swap with an asking price of 1.00 DJED for ADA.  Listing swaps on-chain like this creates extra liquidity.  Arbitragers will constantly look for profitable trades and whenever they see a slight overlap they will know they can bundle all these overlapping swaps together for a small profit.

This is exactly how the order book works in traditional finance except that each user also has the option of being the market maker (arbitrager) themselves.  For example, Mike might list his 49.8 ADA -> 50 DJED swap on-chain and Charlie will be able to see this.  Charlie can choose to be a market maker and bundle up Mike's swap with his own requirement for 50 Ada.  Being a market maker will be easy enough that anyone can do it.  Charlie can take the 50 DJED from Mike's swap, contribute another 50 DJED of his own, and execute Alice's swap using the 100 DJED to receive 100 ADA of which 49.8 will go to completing Mike's swap.  Charlie now has 50.2 ADA less the transaction fee of 0.2 ADA for a net of 50 ADA.

You might think that this will create a lot of competition between arbitragers but it is likely that most users will simply look at the prevailing price of the intended swap and set their price accordingly.  If they want their trade executed quickly then they will set their price so that there is some profit motive for an arbitrager.  Professional arbitragers will run sophisticated searching algorithms looking for long chains of profitable swaps.  In some cases this will mean that the combined bundle of a long chain of swaps might be overall profitable even though some individual links in the chain might not be profitable in isolation.  In other words, it will be cheaper and easier for users to let more sophisticated arbitragers seek out and bundle up lots of swaps together.  Users can just set their price and let the arbitragers look for the cheapest way to combine their swap.

The good thing about this is that Alice, Mike and Charlie need only set the exchange price they are happy receiving and their swaps will get executed when they are 'in the money" after taking into account the transaction fee.  However, unlike the traditional finance world, they each preserve the option of being the market maker (arbitrager) themselves.

> ## FAQ

Reviewer: >  He only needs 98.6 ADA

I know you mean 99.6 but the example still works very well 🤓 ... and I agree no reason to be advertising this arrangement as a limiting factor, because this is seen so extensively use today in
1. offering / asking a more fetching price for larger orders in an exchange crypto order book
2. offering a higher Bitcoin transaction fee to ensure a Tx is put into an early block

> ## FAQ

Author: Your examples seem too technical/mathematical. To make the FAQ as approachable as possible, I think examples should be simplified to the bare minimum.

> I still think you are underselling Cardano-Swaps.

I don't see much of a difference between your version and the current one. Which parts do you think the current version is underselling?

>  no reason to be advertising this arrangement as a limiting factor

Which parts make it come across as a limiting factor? I am personally not interpreting it as a limiting factor and, to me, the current version does not really read that way. I think this paragraph:

> Yes, this is a concurrency bottleneck but no, this is not an issue. This is the perfect niche for businesses to build on top of Cardano-Swaps. Historically, whenever there is a risk that users don't want to carry, a business can form to carry that risk instead, for a price. Insurance is one such example. The same scenario applies here.

can possibly be read in a negative light. I can remove it entirely (aside from the first sentence). Then this paragraph:

> Users can think of this approach as creating a limit order just below the market price so that it is immediately executed; the further the order is set below the market price, the faster it will be executed. Users can decide for themselves how much they are willing to pay for a quick execution and set their prices accordingly.

can be reworded to express how this practice is already very common like in these two examples:

> 1. offering / asking a more fetching price for larger orders in an exchange crypto order book
> 2.  offering a higher Bitcoin transaction fee to ensure a Tx is put into an early block

and I would also add tiered fees suggested for cardano.

> ## FAQ

Author: What do you think are the most important points this answer should convey?

To me, they are:
1. Users can pay to have businesses compete for UTxOs on their behalf. Users can set their own prices for this.
2. Users always have the option to interact with UTxOs directly if desired.
3. Anybody can be a liquidity provider and anybody can get paid to compete for UTxOs. It is fully permissionless. This competition also means users don't need to pay that much to have their swaps executed.

> ## FAQ

Reviewer 3: Yes, sorry, 99.6.

I really think the reader needs to better understand just how good this design is in term of completely decentralising the role of arbitrage seeker.  Moreover, people need to understand that these arbitrage seekers will not be like the MEV bots on Ethereum that front-run users to steal from them.  Instead Cardano-Swaps makes it easy for anyone to be an arbitrage seeker and these bots will continually look for opportunities to bundle up sequences of swaps that result in a closed loop which is even minimally profitable.

And, why wouldn't lots of people run arbitrage seeker bots since even just a penny here or there is still risk free profit.  There is no disincentive to bundle up and execute a sequence of swaps that delivers even a tiny net fraction of some token after the transaction fee is paid.  The deterministic eUTxO model of Cardano means that the arbitrager can't lose.  Either the entire transaction goes through giving him his small fraction of extra tokens or it doesn't because someone else sniped the bundle first.  These arbitrage seekers are not "market makers" in the TradFi sense because they are not taking any inventory risk if they are simply closing out a circular chain of swaps.

So Cardano-Swaps gives users are choice:  They have the option to create a swap and then execute it themselves in the simplistic sense by paying the full transaction fee and bundling only the minimum swaps in order to fulfil their needs.  Or, they can simply submit their swap on-chain and let an arbitrager combine their swap with a number of others and thereby save some transaction fees depending on how aggressively they set their swap price.

I believe these arbitrage seekers are going to get very professional and some will even act as market makers in the TradFi sense because some will be prepared to hold an inventory of tokens and accept directional risk.  This will result in even cheaper costs for ordinary users because they will be able to set their swap prices more aggressively (ie. with less arbitrage profit).

Another difference with TradFi is that Cardano-Swaps makes all pricing, and all fees, transparent.  There is none of this selling of privileged access to the order book because everyone can see the order book on-chain.

Eventually, I can see how Cardano-Swaps could be combined with Hydra for increased throughput.  Imagine a stock exchange running a number of hydra heads with each broker having it's own hydra heads.  Clients can transfer their stock tokens to smart contracts managed by their broker's hydra heads.  The clients can trade on the stock exchange for minimal fees with each broker running bots that bundle up chains of Cardano-Swaps, even involving sequences of different trading pairs for enhanced liquidity.  When the client wants to cash out or stop trading for a while they can simply withdraw their tokens from the broker's Hydra head smart contracts into their own wallet for safe self custody.  This will absolutely crash the infrastructure costs of being a broker and remove the barriers to competition.  Furthermore, users will then demand the transparency guaranteed by these new brokers.  Since Hydra offers isomorphic compatibility with base layer Cardano, the Cardano-Swaps smart contracts can be used unchanged.

> ## FAQ

Reviewer 3: > What do you think are the most important points this answer should convey?
> 
> To me, they are:
> 
> 1. Users can pay to have businesses compete for UTxOs on their behalf. Users can set their own prices for this.
> 2. Users always have the option to interact with UTxOs directly if desired.
> 3. Anybody can be a liquidity provider and anybody can get paid to compete for UTxOs. It is fully permissionless. This competition also means users don't need to pay that much to have their swaps executed.

Yes, I think so.

I also agree that you want to keep the answers as simple as possible.  It is your document, I am really just trying to provide a different perspective to stimulate your thought if that can help paint it in the best light.  All these other DEX designs are inferior and it seems that AXO will not be permissionless.  At the end of the day, I just want to see Cardano-Swaps go for number one!

> ## FAQ

Author: I definitely appreciate the perspectives so don't hesitate to keep sharing!

Your example has a slight issue:

> For example, Mike and Charlie could both create 49.8 ADA -> 50 DJED swaps.

If Alice's swap is 1:1, 49.8 ADA can only get you 49.8 DJED. Mike's and Charlie's swaps would actually have to be 50.2 ADA -> 50 DJED. So the arbitrager would have from the two swaps 100.4 ADA but only needs 100 ADA to get the 100 DJED to satisfy them. The extra 0.4 ADA is left for the arbitrager.

How's this wording using your example (I think your example makes it a little clearer):

> ### If there is one two-way swap UTxO with 1000 DJED and there are 10 users trying to use 100 DJED each, isn't this still a concurrency issue despite having enough liquidity?
>
> Yes, this is a concurrency bottleneck but no, this is not an issue. *Cardano-Swaps addresses these bottlenecks by natively supporting a healthy fee market that is immune to MEV and front-running.* 
>
> Imagine a slightly simpler example: Alice has a 100 DJED swap available for ADA at 1:1 and Mike and Charlie both want DJED from it. No liquidity bottleneck here, just a concurrency bottleneck. If Mike and Charlie do not want to deal with the concurrency bottleneck, they can pay arbitragers to execute the swaps for them. Here is how it would work:
> - Both Mike and Charlie create one-way swaps for 50.2 ADA -> 50 DJED (a price of about 0.99 DJED per ADA). If executed against Alice's swap, Mike and Charlie would each get 50 DJED and whoever executes their swaps gets to keep the other 0.2 ADA per swap. These swaps are effectively saying: "I'll pay you 0.2 ADA to execute this swap for me."                                                                                                                                                                          
> - Sarah, an arbitrager, can see all three swaps and executes them, keeping the 0.4 ADA arbitrage (0.2 ADA from Mike's swap and 0.2 ADA from Charlie's swap). After the 0.2 ADA transaction fee, she pockets the remaining 0.2 ADA. 
>
> In other words, Mike and Charlie deliberately create an arbitrage opportunity so that arbitragers will want to execute their swaps. If Charlie really needs his swap executed ASAP, he can configure his swap so that he gets 50 DJED and the arbitrager gets 0.5 ADA for executing his swap. Basically, the more attractive the buyer makes the arbitrage opportunity, the more arbitragers will compete for it and the faster their swap will be converted. 
>
> From the perspective of the arbitragers, the only cost to them is the transaction fee which scales very slowly relative to the number of swaps in the transaction (14 swaps costs about 1.5 ADA). Combine this with the fact that these transactions are risk free for arbitragers (a failed transaction is no cost to the arbitrager) and you end up with very healthy incentives for arbitragers to *want* to do this and satisfy as many users as possible with each transaction. The risk free nature means arbitragers would even be willing to do this for small profits like 0.2 ADA. Being an arbitrager is fully permissionless so anyone can start generating income this way. 
>
> These incentives are how Cardano-Swaps natively creates a fee market. It is effectively no different than paying a larger transaction fee to ensure your transaction gets into the next block. Furthermore, since users must explicitly set their prices which are trustlessly enforced, MEV and front-running are simply not possible. 
>
> As this scenario shows, most users will likely not be competing for swaps - arbitragers and other businesses will be. Most users will create orders via swap UTxOs that are then swept by these businesses. This results in there being zero concurrency problem for the end user. 
>
> When it comes to creating a decentralized economy, not every feature needs to be directly enabled by                                                                                                               
the DApp. Some features can, and should, be left to the market to fill.

I want the answer to be fairly short but still address the key points. To me the key take-away is that Cardano-Swaps has an endogenous fee market and that is how it solves the concurrency bottleneck. This fee market is immune to MEV and front-running (I think... I have not really thought this through yet.).

> ## FAQ

Reviewer 3: Feel free to make up some more FAQs if you like.  For example, you could add an FAQ about using Cardano-Swaps with Hydra.

> ## FAQ

Author: I definitely want to mention hydra at some point but right now, with hydra not really finalized, there is too much speculation for that. Once I'm done with the protocols, I do want to try playing around with hydra and see what I can do with it. For now, I want to keep the FAQ to the criticisms I see most often. Since I also want to release it this weekend, it is a bit late to add much more...

> ## FAQ

Reviewer 3: @fallen-icarus : That re-write above is awesome, it really showcases the strengths.  In the end, I think the arbitragers will be so actively looking for any profitable bundle of swaps that users will simply post swaps that are very minimally in the money if they need them quickly executed.  And, in reality any tiny profit margin is still risk free profit.  Even it is only a penny worth, as an arbitrager, you already have the software running so why not take that penny.

# Cardano-Swaps

A [p2p-DeFi protocol](https://github.com/zhekson1/CSL-DeFi-Protocols) for trustlessly swapping
fungible tokens on the Cardano Settlement Layer. All users maintain full custody, delegation
control, and voting control of their assets at all times. No batchers are required.

> **Note** Knowledge of basic Haskell syntax and `cardano-cli` usage is recommended. For a list of
> everything that has changed from the previous version, see the [Changelog](CHANGELOG.md).

The Getting Started instructions can be found [here](./GettingStarted.md) and the benchmarks can
be found [here](./Benchmarks/).

---
## Table of Contents 
- [Abstract](#abstract)
- [Motivation](#motivation)
- [Cardano-Swaps](#the-cardano-swaps-protocol)
- [Specification](#specification)
- [Benchmarks and Fee Estimations](#benchmarks-and-fee-estimations-ymmv)
- [Features Discussion](#features-discussion)
- [FAQ](#faq)
- [Conclusion](#conclusion)

## Abstract

Cardano-Swaps is a p2p-DeFi protocol for swapping fungible tokens on the Cardano Settlement Layer
(CSL). It solves many of the pitfalls of current DEX implementations by empowering users to deploy

> their own (and interact with each others') script addresses. This leads to the formation of an

Reviewer 2: ```suggestion
their own (and interact with each others') script addresses. This enables the formation of an
```

order-book style "distributed" DEX. The protocol supports two kinds of swaps: one-way swaps and
two-way swaps. One-way swaps emulate the basic limit order. Meanwhile, two-way swaps provide a
mechanism for naturally incentivizing users to provide liquidity to the DEX without having to rely
on yield farming. Furthermore, since swaps can be freely composed, arbitragers are naturally
incentivized to spread the available liquidity across all trading pairs. This protocol is natively
composable with all DApps.

## Motivation

Many DEXs on Cardano are currently implemented in ways that lock users' assets into a tightly fixed,
and/or centrally maintained, set of script addresses and UTxOs. This *concentrated DApp* design
pattern is reminiscent of Ethereum's accounts-based programming paradigm.

> However, the concentrated DApp design does not map well to eUTxO based blockchains. To be exact, a

Reviewer 2: ```suggestion
However, the concentrated DApp design does not map well to eUTxO based blockchains. To be precise, a
```

of a trustless and decentralized economy, especially those built on Proof-of-Stake (PoS)
blockchains.

##### Why the LP based architecture is not the right architecture for DEXs (in no particular order)

- No matter how performant a system of batchers is, their resources do *not* scale in proportion to
the number of users unless new batchers can permissionlessly join when demand is high. 
- Since an economy's health is proportional to how easily and accurately price discovery can occur,
DEXs that do not allow users to freely express their own desired prices can only result in an
unhealthy blockchain economy. Furthermore, impermanent loss (a consequence of not being able to
express desired prices) makes it impossible for users to properly manage risk.
- Batcher based protocols lack trustless composability with other dApps; a critical feature for both
scaling eUTxO-based DeFi and building a fully featured, trustless blockchain economy.
- The concentration of assets into a single address or few addresses directly undermines the
security assumptions of PoS blockchains. This ultimately means it is *impossible* for LPs to serve
as the foundation of a trustless and decentralized economy.

> 

Reviewer 3: Perhaps add:
The LP based architecture pools user assets at an address outside of their direct control.  This forces users to choose between using their assets to provide trading liquidity or being able to harness the voting power their assets provide.

Author: What if I added this bullet to the end:

> - The concentration of assets forces uses to give up voting rights over their assets. In essence, users are forced to pick between participating in DeFi and participating in decentralized governance. Furthermore, whichever entity controls the voting power of the concentrated assets has tremendous influence over the decentralized government.

Reviewer 3: I think this point is so important because it not only limits user voting power, but it also fundamentally undermines Cardano's security.  I like what you have said already, but I somehow want it to be harder hitting.

Author: What about:

> - The concentration of assets into a single address or few addresses directly undermines the
security assumptions of PoS blockchains.
> - The concentration of assets forces users to give up voting rights over their assets. In essence, users are forced to pick between participating in DeFi and participating in decentralized governance. Furthermore, whichever entity controls the voting power of the concentrated assets has tremendous influence over the decentralized government.
> 
> The last two bullets especially mean it is *impossible* for LPs to serve as the foundation of a trustless and decentralized economy.

I think the final sentence is true and emphasizes it well. I'm not sure how to emphasize it better than that.

Reviewer 3: The current DEX design is quite simply fundamentally flawed for a POS blockchain.  Consequently the DEX owners come up with these "governance" tokens and "staking decision" tokens.

Cardano-swaps design is beautiful by comparison.

in a transaction).
- No delegation control or voting control while using the DEX - DEXs try to issue governance tokens
to get around this but fair distribution of the governance tokens is rarely accomplished (it is a
very hard problem). Even if the governance tokens were distributed fairly, they are largely just for
show since the DEX creators can still choose to go against the vote. There is no trustless
connection between the governance tokens and what actually happens with the stake/dApp. Finally,
even if there was a trustless connection, there is no way to actually have all users express their
own stake preferences when the stake is decided democratically. It is almost inevitable that the
voting outcome will go against the wishes of some users. *This is in direct violation of the
security assumptions of PoS.* No matter how you look at it, LP based dApps are an existenstial
problem for PoS blockchains. As LP based dApps become more adopted, the underlying PoS blockchain
will only become more centralized. LPs can never be used as the foundation for a trustless and
decentralized PoS blockchain economy.

> 

Reviewer 3: Maybe add:
A whale that contributes 51% of the liquidity pool assets will get to decide the voting direction.

Author: This seems difficult to concisely add since it has a lot of implicit concepts. For example, a whale that controls 51% of staked ada can decide the direction of the blockchain. Why is it okay when talking about Cardano in general but not okay when talking about the DApp? The problem is in how likely it is to acquire the 51% but now I would need to explain the difference. I am honestly not sure how to best articulate this difference. I was hoping my sentence "fair distribution of governance tokens is rarely accomplished..." was enough to allude to control by a whale. Do you really think it is not enough?

Reviewer 3: What you have already I think is OK.  I just wasn't sure if people really get the significance in general.

A whale can actually leverage his voting power by participating in these liquidity pools where majority decides using a governance token.  For example, lets say I have 500M Ada and there are a couple of liquidity pools each with 200M Ada in them.  I can now contribute an extra 201M Ada to each of these liquidity pools so that each pool now has 401M Ada but because I have more than 50% of each pool then I get to choose the voting direction.  I now have control over 802M Ada in voting power plus the extra 98M Ada I still have in my wallet.

There are other scenarios where I could leverage my voting power even more.  For example, I can watch the voting of a particular liquidity pool and then contribute just enough tokens to get the vote over 50% in the direction I wanted, which may require much less tokens than 50% depending on how others in the pool are voting.

Imagine I used this technique to control how Ada was staked to particular pools in my malicious "coalition":  I could provide enough Ada to various liquidity pools and then decide how this Ada was staked and I get the extra Ada from the other smaller liquidity providers in addition to my own.  My coalition of pools thereby gets more Ada staked with it and therefore more control over the block production.

This pooling of assets in liquidity pools is a very bad idea because it hands more power to whales that possess enough capacity to shift the balance.  I don't think most people realise that liquidity pools with governance tokens allow whales to actually increase their power.

Your design fixes this huge threat to control of Cardano's block chain.

Author: > I just wasn't sure if people really get the significance in general.

I think this is a scenario where actions speak louder than words. Speculators simply don't care and developers are incentivized not to care since liquidity pools are "what's hot" and where the money is. My design has literally no way for me to profit as the creator which also hurts the chances of developers coming around. Not having an insider token also makes it very difficult to get funded. "It is difficult to get a man to understand something when his salary depends upon his not understanding it." I think (hope) those that are open to listen should be easily swayed by what has already been said.

TBH I think the situation is actually self-correcting. All of the governance issues of liquidity pools are true, but the more immediate concern is that they are economically unstable and will likely collapse within the next several years. Yield farming is just hyper-inflation waiting to happen. Most LPs say their yield farming is temporary but it fundamentally cannot be temporary. How will liquidity be incentivized without it? Transaction fees are *not* paid in the token which means there is no refilling pot like with Cardano's reserves (ideally). Where will the new tokens required to pay liquidity providers come from? These tokens are economically useless as well so the idea that the DApp can provide a service to raise a reserve pot is also extremely unlikely. The only option left is to mint more. 

I think the only reason LPs haven't collapsed yet is because users are staking the yield tokens to earn more yield tokens. Once users decide to "cash out" en masse, like during the beginning of a bear market, the weaknesses will be exposed and I think a lot of users will get hurt. The beginning of a bear market is also when the risk of impermanent loss will be highest which further drives users to sell the yield tokens in order to recoup their losses. I don't see how this doesn't result in the value of the tokens very quickly going to zero. Without any incentive to take on the risk of impermanent loss, users will run for the exits. It would be a run on liquidity pools. I wouldn't be surprised if this scenario resulted in a lot of lawsuits and criminal charges against the developers.

I don't see how the distributed design doesn't win by default.

Reviewer 3: Yep.  I have never participated in any of this liquidity pools - yield farming craziness for exactly the reasons you have articulated.  I think they are worse than Ponzis because aside from hurting the direct users, they can also cause collateral damage to other parts of DeFi through re-hypothecation and other dapps building upon or using governance tokens as collateral.  Furthermore, they might  even hurt the proper functioning of the base protocol due to the control over how their pooled Ada is staked.  They are bad design using dreamed up additional tokens to create Ponzi-like artificial demand.

Cardano-Swaps is an order-book based DEX that takes inspiration from Axo's programmable swaps
[design](https://www.axo.trade/whitepaper.pdf), adds delegation control as a foundational feature,
naturally incentives its own liquidity, and, through the use of Beacon Tokens, removes the need for
specialized indexers. In short, it solves all of the pressing issues that made order-book DEXs
inferior to LP based DEXs. 

### High-Level Overview

The protocol is comprised of two types of composable swaps: a one-way swap and a two-way swap.
One-way swaps are effectively just limit orders (ie, they can only be executed at the specified
price or better). Most users will likely use the one-way swap.

Two-way swaps are a mechanism for profitably providing liquidity to the protocol without having to
resort to yield farming. These two-way swaps can go in either direction (eg, a two-way ADA <-> ERGO
swap can go ADA -> ERGO or ERGO -> ADA). By allowing the liquidity provider to specify separate
prices for each direction, they can profit from either direction of the swap. For example, if the
market rate for DJED <-> USDC is 1:1, Alice can set DJED -> USDC to 1.01 USDC per DJED and USDC ->

> DJED to 1.01 DJED per USDC. This means Alice makes a 1% return no matter what direction is used.

Author: >  In other words, a permanent price change would not protect the two-way swap user from losses?

I'm not sure I follow. No loss has occurred in your example (assuming the exchange rate is 1:1): before the swap the consolidation, the total value is $200; after the consolidation, the total value is $201. That is a profit. The fact that the price of DJED in USDC remained above 1 is not an issue, imo. 

Two-way swaps are meant to be a "last resort" swap because you will always have to pay a slight premium, no matter the swap direction. But that is not the same as saying there is no economic utility for this feature. If Alice has DJED but really needs USDC, she will use whatever she can to get the USDC. I think most users will be okay paying the slight premiums in this scenario. Since two-way swap providers are competing with each other, the premiums will approach the true value. Basically, if Alice really needs USDC, she has a choice of missing the opportunity that requires USDC or paying up marginally. 

Consider this, with all the stablecoins being created, liquidity is fractured into "silos" (one for each stablecoin). Two-ways swaps makes the liquidity of stablecoins freely convertible at a slight premium. I think there is tremendous utility for this feature, especially since the conversion can happen in the same transaction as other p2p protocol executions (like buying an options contract that requires DJED). Personally, if I have DJED but need USDC to buy an options contract, I would be perfectly fine paying a 1% premium to acquire the USDC (assuming the options contract can recoup the expense).

> I do like the idea of this two-way swap for things like stable coins, where the price is expected to fluctuate around parity...

This is the number one use case for two-way swaps. They should be seen as active trades, not "set and forget" trades. You are correct that they are more suited for non-volatile pairs but there could still be value in active traders using them for volatile pairs, although this then carries a higher risk of loss.

The way I see it, two-way swaps incentivize providing liquidity to main trading pairs such as stablecoin <-> stablecoin or ADA <-> stablecoin. Then, the arbitragers take this liquidity and spread it. There do not need to be two-way swaps for all trading pairs for their to be liquidity on the DEX. Only the entry and exit points (the main trading pairs) need liquidity.

> DJED to 1.01 DJED per USDC. This means Alice makes a 1% return no matter what direction is used.

Reviewer 3: > I'm not sure I follow. No loss has occurred in your example (assuming the exchange rate is 1:1):

That is because we are assuming the stable coins will remain at 1:1 exchange rate.  If one stable coin permanently de-pegs then the owner of the two-way swap will end up holding the de-pegged coin only and will therefore suffer loss if the price never re-peggs.  Similarly if you create a two-way swap for any other asset pair, then you will end up holding only the asset which becomes relatively less valuable.  For example, if you create an ADA <-> DJED two-way swap at 1 ADA -> 1.01 DJED and 1 DJED -> 1.01 Ada, then if Ada goes above 1.01 DJED then your swap contract will now only hold DJED because arbitragers will fully execute it in the DJED -> ADA direction.  In other words as the price of Ada increases you will be left holding only the less valuable asset.

The way I see the two-way swap is a bit like an individual AMM swap but instead of having an exchange rate that varies in an attempt to offset the relative price change (eg. using constant product formula), instead there is like a two-way price cliff where once the price is exceeded in one direction then the contract will get fully executed in the direction of the weaker asset.

I like the concept for stable coins where they are expected to fluctuate around parity with each other, but I can't see it working well for any other asset pairs.  I don't think it is fundamentally unsound or anything.  I just wouldn't use a two-way swap myself for anything other than a stable coin pair and only if I was extremely confident in each coin ultimately returning to peg.

For stable coins, I do agree this is likely to improve liquidity because I believe people will want to create such swaps in order to accumulate stable coins as the price fluctuates either side of parity.

> DJED to 1.01 DJED per USDC. This means Alice makes a 1% return no matter what direction is used.

Author: > ... there is like a two-way price cliff where once the price is exceeded in one direction then the contract will get fully executed in the direction of the weaker asset.

This is why it is still an active trade. Stablecoins are the safest use-case (as you said, a de-peg is still a risk) but I do think some users would make two-way swaps for larger assets <-> stablecoin, like ADA <-> DJED. I'd personally be interested in it during sideways market movements and manage my risk by limiting the amount of assets in the ADA <-> DJED swap and charging a very high premium in the direction I think is weakest. I would expect the vast majority of two-way swaps are for stablecoin <-> stablecoin but I do think there will be at least *some* two-way swaps for other pairs. Professional traders can charge higher premiums for these since they have less competition. I think it is an open question as to whether there will be *enough* two-way swaps for the other pairs but I think there is plenty of incentive for it. Arbitragers themselves may be willing to pay premiums as high as 30% for a swap like ADA <-> DJED if the arbitrage opportunities are profitable enough. Considering the very nature of illiquidity, I think there actually will be enough profit potential for arbitragers to accept the 30% premium. 

I will concede that this is speculation on my part, though. I am personally interested in building an automated arbitrager and imagining what I would do.

> DJED to 1.01 DJED per USDC. This means Alice makes a 1% return no matter what direction is used.

Reviewer 3: I can see how you could use the two-way swaps as part of an automated arbitrager with it updating the exchange rates as the price changes.

> DJED to 1.01 DJED per USDC. This means Alice makes a 1% return no matter what direction is used.

Author: Thinking about this yesterday made me realize there is a hole in the incentives that I need to patch before I release the beta. This hole mainly impacts swaps like ADA <-> stablecoin. I don't think it really impacts stablecoin <-> stablecoin. Basically, the chaining of swaps opens up the possibility of a flash swap (name deliberately similar to flash loan) where a swap's assets are taken and returned in the same transaction. Since the swap's balance didn't change, the protocol will not charge the premium. I'm going to explain it here, partly to make sure I understand it fully myself and partly to see if anyone has feedback on it.

The way I was initially imagining arbitrage was like this: `ADA -> x -> y -> z -> ADA`. Each arrow is a swap. In this scenario, the start and end swaps are different swaps. But it is actually possible to do this: `ADA -> x -> y -> z -> x -> ADA`. In this scenario, the start and end swaps can be different swaps but don't have to be. When `z -> x` is executed, the resulting `x` asset can be returned to the `ADA -> x` swap and the ada used in that swap can be reclaimed. So on net, the `ADA -> x` swap had no change in the balance of the swap UTxO. This swap is a flash swap: `x` is taken and returned in the same transaction. Since the balance of the swap did not change, the swap does not require any payment. I have manually confirmed that this is indeed possible.

This technically does not hurt anyone since no assets were officially taken without paying. The issue is the arbitrager used the flash swap to make a profit but did not pay the user for providing the liquidity. Since no premium needs to be paid to do this, there is a massive incentive for arbitragers to use flash swaps. Since stablecoin <-> stablecoin swaps are unlikely to be chained, I don't think this really impacts them. This definitely impacts all other swaps, though. While I think there are more use cases for providing liquidity for other trading pairs than just arbitrage, providing liquidity for arbitragers is the main incentive since they will likely be the most active. So the arbitrager being able to use the swap's liquidity without paying severely dampens the incentive for users to provide liquidity.

Considering what smart contracts can see about the transaction and given the redundant executions, there is no cheap way to tell how much of the flash swap's assets were used and charge the premium accordingly. But this actually may not be an issue. The premium is meant for scenarios where the swap's assets were actually converted. Since no assets were actually taken, I do not think the premium needs to be charged. Instead, a flat fee for using the swap seems more appropriate.

At a high level, if I added a `minimumGive` field to the datums, users can specify how much they want of the ask asset per execution. This would apply both to normal swaps and flash swaps. For a normal swap, if Alice has 10 ada in a `ADA -> DJED` swap and specifies a `minimumGive` of 2 djed at a 1:1 ratio, if Bob wants to swap with Alice, he must give at least 2 djed in the swap. It is still possible for Bob to take any amount of ada up to the ratio, but taking less than 2 ada would be a loss for Bob. Bob is incentivized *not* to do this.

For flash swaps, it does not matter that none of the swap's offer asset was taken; the `minimumGive` will still be  enforced. In this scenario, this `minimumGive` is the fee the liquidity provider is charging for allowing the arbitrager to use their swap as a flash swap. If the swap is actually converted, the premium will still be enforced. Considering arbitragers will prefer this fixed fee over the premium, I think this actually makes it safer for users to provide liquidity of larger trading pairs - most swaps will likely be arbitragers using the swap as a flash swap.

I had considered a `minimumTake` field instead but the quantity of the offer asset is what is important to arbitragers. If the quantity must decrease which each execution, eventually the swap would be unusable by the arbitragers. The `minimumGive` allows the swap to be used indefinitely while still having the same result.

> DJED to 1.01 DJED per USDC. This means Alice makes a 1% return no matter what direction is used.

Author: I've been testing this more and I may be wrong about the flash swap being useful here. Intuitively, since the flash swap doesn't change, it shouldn't be needed. And it turns out, it is not needed. The above example of `ADA -> x -> y -> z -> x -> ADA` can actually happen without the `ADA -> x` swap. It would effectively be this:

```mermaid
graph TD;
    X-->Y;
    Y-->Z;
    Z-->X;
```
The `x` that is needed to unlock the `y` in the first swap comes from the last swap. It is circular. So the inclusion of the `ADA -> x` swap does nothing but increase the transaction fee. There may actually be no need for the `minimumGive` but I need to think about this more.

> DJED to 1.01 DJED per USDC. This means Alice makes a 1% return no matter what direction is used.

Reviewer: In light of this, it does sound like the protocol parameters from which the minimum Tx fees are derived, having been chosen primarily to prevent DDoS on the network, would also cover this similar case.

> DJED to 1.01 DJED per USDC. This means Alice makes a 1% return no matter what direction is used.

Reviewer 3: For what it is worth, I agree with your reasoning here and agree that the extra complexity is not needed, and would make your design less neat.

> DJED to 1.01 DJED per USDC. This means Alice makes a 1% return no matter what direction is used.

Author: I can't come up with any situation where flash swaps would be useful. Any time you could use a flash swap, the circular swap would be cheaper. So I don't see a need to add the `minimumGive`.

DJED), but it also opens up the possibility for arbitragers to find profitable arbitrage
opportunities. This potential for profit naturally incentivizes liquidity to be spread across all
trading pairs instead of it being siloed into one trading pair as in TradFi and LP based DEXs.

Finally, through the use of beacon tokens, the protocol is fully p2p. No batchers are required
although they can be used if desired. And thanks to beacon tokens, users can get their own addresses
which guarantees users maintain full delegation and voting control of their assets at all times.

The only remaining challenge is the querying capacity of existing off-chain APIs, such as Blockfrost
or Koios. (This is not a limitation for users with powerful enough hardware, as they can run their
own API database).

## Specification

Each type of swap is comprised of a validator / beacon policy pair. This means that each type of
swap gets its own universal address. For example, all of Alice's one-way swaps, regardless of the
trading pair, will be located at address X and all all of Alice's two-way swaps, regardless of the

> trading pair, will be located at address Y. Address X is != address Y. Furthermore, since each type

Reviewer: good catch 🤓 (the existing text "reads" properly as language, but violates programming convention)

of swap has its own beacon policy, each type is easily distinguishable off-chain while still being

> easily queryable by end users.

Reviewer 3: Is this restriction necessary?  Why can't every swap be at a different address since the beacon tokens will still label the eUTxOs to make them easily searchable?

Author: It is not necessary but I thought managing a single address would be a better UX and make it easier for front-ends to integrate it. I don't want scenarios where users create a swap for an obscure trading pair and then forget about it. If it is in a different address, they may forget about it entirely. But if all swaps are in a single address, the user will see the swap the next time they check their DEX address. I personally don't know of any use cases that require separate addresses. If you really want a separate address, you can just use a separate staking credential.

Reviewer 3: I see your point.  I thought there might be a way for your wallet + smart contract app to calculate additional addresses in a similar way to how wallets do this now by generating additional payment keys from one master key.

I am not sure of the pros and cons apart from my concern re quantum resistance and previously exposed public keys rather than just a hash of the public key.  But maybe this doesn't apply to smart contract public keys and hashes of public keys.  (See my other comment below related to this.)

I was mainly adding this comment to get you thinking because you obviously know more than me about such things.  I understand that quantum resistance is the main advantage wallets gain from generating additional addresses.

This modular approach of having a separate script pair for each swap makes it easy to add new kinds
of swaps in the future.

:exclamation: While the spec shows one-way swap datums and redeemers having the same names as the
two-way swap datums and redeemers *they are distinct data types*. Imagine they are in separate name
spaces.

### One-Sway Swaps

One-way swaps use one universal validator script for all swap pairs, and one universal minting
policy for all one-way beacons. There are two beacons for one-way swaps: the trading pair beacon and
the offer beacon.

The offer beacon asset name is `sha2_256( offer_policy_id ++ offer_asset_name )`.

The trading pair beacon asset name is 
`sha2_256( offer_id ++ offer_name ++ ask_id ++ ask_name )`. However, in the case where one
of the assets is ADA (which is just the empty bytestring), the ADA policy id is set to "00" instead.
Without this changed, there would be no difference between the offer beacon and trading pair beacon

> when ADA is part of a swap.

Reviewer 3: Maybe an example?  Would ADA->DJED look like this?:
sha2_256( "00" ++ "ADA" ++ "1234" ++ "DJED")

Author: No, it would be: `sha2_256( "00" ++ "" ++ "1234" ++ "DJED")`
For one-way swaps, the other direction would be: `sha2_256( "1234" ++ "DJED" ++ "00" ++ "")`

I can add an example to the readme.

These beacons allow for expressive queries since it makes it possible to query all swaps where XYZ
is being offered, regardless of the asset being asked for. This is in addition to being able to
query all swaps for a specific trading pair. This extra expressiveness improves the arbitrager's

> ability to find profitable paths through currently open swaps.

Reviewer 3: When searching there is only 1 offer_policy_id and you know the offer_asset_name you want so you calculate the sha2_256 sum and search for this hash value which represents the offer beacon?  Are you using the sha2_256 sum to limit the beacon name size rather than simply using the offer_policy_id + offer_asset_name directly?

Author: > Are you using the sha2_256 sum to limit the beacon name size rather than simply using the offer_policy_id + offer_asset_name directly?

That is correct. Token names are capped at 64 characters and policy ids take up 56 of the 64 characters so simply concatenating them is not possible.

##### Swap Address

All users get a single swap address where all of their one-way swaps will be located. This universal
validator can enforce all possible swaps for the user. And since users only need to worry about a

> single address, it keeps the UI simple.

Author: > One of the reasons Cardano wallets choose new addresses is to provide quantum resistance because only the hash of the public key is known until the first eUTxO is spent from that address.

I did not know this. Can choosing different staking credentials be enough to provide the same resistance? You can always get a different address if you use a different staking credential.

> single address, it keeps the UI simple.

Reviewer 3: No, as I understand different staking credentials won't help with quantum resistance because you will have already executed a transaction that required exposing your public key when you stake with a pool.

I could have this a bit wrong, but here is how I understand this quantum resistance feature to work.  Normal addresses on Cardano are actually the hash of the public key.  Since the hash used is quantum resistant, this means that the public key is not actually known until a transaction is actually executed for that address.  When a transaction is executed, you provide a copy of your public key + signature using private key and these are checked firstly to confirm that the signature is correct and then also that the public key hash which represents the address is correct.

This is why wallets like Yoroi, daedalus and others generate a new address each time when you click on the receive funds tab.  The designers were tricky in using the hash of the public key as the address rather than just the public key itself.  The hash used is apparently quantum resistant, but ed25519 keys are supposedly not.

> single address, it keeps the UI simple.

Reviewer 3: Yes, the quantum resistance effect by creating new addresses which are the hash of an unused public key is a nice trick but it is very clear that we will need updated cryptology when quantum computers get closer.

data SwapRedeemer
  = CloseOrUpdate
  | Swap
```

Only the owner (signified by the address' staking credential) can use the `CloseOrUpdate` redeemer.
Anyone can use the `Swap` redeemer as long as the swap conditions are met.

##### Datums
Inline datums are used for all UTxOs. All UTxOs get the same type of datum:

``` Haskell
data SwapDatum = SwapDatum
  { beaconId :: CurrencySymbol -- ^ Beacon policy id for one-way swaps.
  , pairBeacon :: TokenName -- ^ Trading pair beacon asset name for this swap.
  , offerId :: CurrencySymbol -- ^ Offer policy id for this swap.
  , offerName :: TokenName -- ^ Offer asset name for this swap.
  , offerBeacon :: TokenName -- ^ Offer beacon asset name for this swap.
  , askId :: CurrencySymbol -- ^ Ask policy id for this swap.
  , askName :: TokenName -- ^ Ask asset name for this swap.
  , swapPrice :: Rational -- ^ The desired swap ratio: Ask/Offer.
  , prevInput :: Maybe TxOutRef -- ^ The output reference for the corresponding swap input.
  }
```

The beacon information in the datum prevents misuse of beacons. The validator forces all beacons to
be burned instead of being withdrawn. This ensures that beacons can only ever be located at

> swap addresses. **If the wrong beacon information is supplied, assets can be locked forever.** For

Author: That specific UTxO will be unspendable. This is the only consequence. This UTxO can never have a beacon token because the datum is misconfigured so other users will never see it when they query the DEX. I can change the wording to point this out.

> swap addresses. **If the wrong beacon information is supplied, assets can be locked forever.** For

Author: I think changing "assets" to "UTxO" should suffice:

> If the wrong beacon information is supplied, the UTxO can be locked forever.

> swap addresses. **If the wrong beacon information is supplied, assets can be locked forever.** For

Reviewer: That specific wording change is helpful, but perhaps due to my own incomplete reading or understanding I don't see how a UTxO still in the user's possession (by retaining the private key for its spending) can become unspendable.  I feel I may be missing something really basic here, but if not then perhaps an explanation of how the UTxO gets "locked" (and what exactly this means) can be put into the document.

> swap addresses. **If the wrong beacon information is supplied, assets can be locked forever.** For

Author: It is a little technical. Basically, the spending script doesn't know what the beacon is by default. You need to tell the spending script which token is the beacon token by using the appropriate fields in the datum. The spending script then requires that those beacons *must be burned* in order to close the swap. Beacons cannot be withdrawn from a DEX address. What would happen if the datum is misused and says ada is the beacon? The spending script doesn't care, it will still demand that the asset set as the beacon in the datum must be burned instead of withdrawn. But ada is not burnable. Therefore, this requirement is unsatisfiable and thus, the UTxO is unspendable.

It is not possible to have beacons stored with misconfigured datums but it is possible to create UTxOs without beacons that do have misconfigured datums. Whether or not the misconfigured UTxO is spendable depends on whether or not the asset set as the beacon is burnable.

> swap addresses. **If the wrong beacon information is supplied, assets can be locked forever.** For

Author: I guess at a high level it can be summed up as: 

Beacons should not be allowed to be withdrawn from a DEX address; it should only be burned. But since the spending script must rely on the datum to know which asset is the beacon, misusing the datum can result in the spending script protecting the wrong asset.

> swap addresses. **If the wrong beacon information is supplied, assets can be locked forever.** For

Reviewer: OK, then unless I've misunderstood, the part I was missing (and is also missing from the above elaboration) must be: the funds to be DEXed at that point are not in your regular assets, but are at UTxO's that you designate for Cardano-Swaps.  These particular addresses (not your regular ones) are governed by spending scripts that might no longer be satisfiable if the Beacon Tokens are used improperly.

If that is true then I think it would help to also include this general point, because I didn't readily understand the problem... and so might other technical but uninitiated readers.

> swap addresses. **If the wrong beacon information is supplied, assets can be locked forever.** For

Author: You are correct in saying that it is the swap UTxO that can be locked (ie, your actual limit order), not your personal funds. Although I don't think it is accurate to say "if the Beacon Tokens are used improperly" since the true beacon tokens can never be misused by design. This locking is only possible when the user doesn't use the true beacon tokens at all. To use the tagging analogy, locking is only possible if a user does not try to tag the Swap UTxO in a way to make it discoverable by others. The act of trying to properly tag it is enough to prevent this locking since the minting of the tag will fail if the datum is wrong and provide you with an error message explaining what is wrong.

Would saying "the Swap UTxO" instead of just "the UTxO" suffice? Explaining it makes it easy to get lost in the weeds and I think most developers can figure it out.

> swap addresses. **If the wrong beacon information is supplied, assets can be locked forever.** For

Reviewer: I am sure most developers can deduce it... since I'm not a developer (more of an integrator, architect, project manager, etc.) I didn't get it right away.  I might have if I'd read the document a few more times but we can't always rely upon that.

There will be other non-developers reading this page mainly to understand the value of implementations, so in this case (and maybe others) it does need to be broadened a bit beyond the audience of seasoned developers to allow all interested parties to immediately understand & qualify the risk. 😇

> swap addresses. **If the wrong beacon information is supplied, assets can be locked forever.** For

Author: I apologize if my comment was slightly offensive, that was not my intention. How's this?

> The beacon information in the datum prevents misuse of beacons. The validator forces all beacons to be burned instead of being withdrawn. This ensures that beacons can only ever be located at swap addresses. **If the wrong beacon information is supplied to the datum, the resulting Swap UTxO can be locked forever.** Simply trying to store the proper beacons in the Swap UTxO is enough to prevent accidental locking due to misconfigured datums. All minting policies check that the datums have the proper information; the mint transaction will fail with the relevant error message if they do not.

> swap addresses. **If the wrong beacon information is supplied, assets can be locked forever.** For

Reviewer: no worries on my account; I didn't take it that way 😎 ... and for the general community I think the explanation above will work well to address that question in advance:
- Less informed readers wouldn't be concerned by the boldface sentence, since it implies that someone would already have to be using the method very deliberately in order for funds to be potentially locked & lost.
- Most of those less technical readers will further conclude that some implementation of Cardano-Swaps will be handling those details & therefore general users wouldn't be at risk by executing the instructions incorrectly themselves. 🤔

> swap addresses. **If the wrong beacon information is supplied, assets can be locked forever.** For

Reviewer 3: > If the wrong beacon information is supplied, assets can be locked forever.
Maybe I am not understanding, but won't the user still be able to withdraw his locked assets from the swap using his staking key?

> swap addresses. **If the wrong beacon information is supplied, assets can be locked forever.** For

Author: No, if the datum is misconfigured, it is possible for the UTxO to be unspendable. The spending script doesn't actually know which asset is the beacon; it relies on the datum to tell it which one is the beacon. It then requires that asset to be burned instead of withdrawn whenever a swap is closed. What if the datum sways ada is the beacon?

> swap addresses. **If the wrong beacon information is supplied, assets can be locked forever.** For

Reviewer 3: It does sound like you have thought about this problem deeply such that your design stops it from happening except if people deliberately act against the design.

However, I am still a little unsure if it is necessary to do the extra checking that the beacon and datum align correctly in the case where the user is withdrawing using his staking key.  If the user can sign the staking key then why not allow the asset(s) to be withdrawn and the beacon gets deleted, without doing the check that the beacon is correct?  That way if the user somehow creates a swp with asset and misconfigured beacon, deliberately or otherwise, then he can still withdraw using the staking key and the misconfigured beacon still gets deleted.

If it is possible to create misconfigured beacons maliciously then this will still require the minimum Ada in the eUTxO, but wouldn't this be better if the user had a way of cleaning them up and getting the 2 Ada min deposit back?  Would such malicious beacons pollute the chain and make querying for valid swaps less efficient?

> swap addresses. **If the wrong beacon information is supplied, assets can be locked forever.** For

Author: Considering how difficult it is to misconfigure a datum, I don't want to add the logic for the ada exception which would hurt the composition for all users. I don't think it is that important to hurt performance.

##### Creating a One-Way Swap

Swap "writers" must first create a swap by using the `CreateSwap` Redeemer in the following manner:

In order to mint beacons with this redeemer, **all of the following must be true**:

1) The beacons must go to an addres protected by the one-way swap validator script.
2) The beacons must go to an address using a valid staking credential.
3) The UTxOs with the beacons must have the proper value:
    - Exactly two kinds of beacons: pair beacon and offer beacon.
    - The beacons must correspond to the beacons in the datum.
    - There must be exactly 1 of each beacon.
    - No extraneous assets are in the UTxO. ADA is always allowed.
4) The beacons must be stored with the proper inline `SwapDatum`:
    - `beaconId` == this policy id.
    - `pairBeacon` == pair beacon asset name for this swap.
    - `offerId` == asset_id of the offer asset.
    - `offerName` == asset_name of the offer asset.
    - `offerBeacon` == offer beacon asset name for this swap's offer asset.
    - `askId` == asset_id of the ask asset.
    - `askName` == asset_name of the ask asset.
    - `swapPrice` denominator > 0
    - `swapPrice` > 0

> 5) The offer asset and ask assets must be different assets.

Reviewer 3: I am just trying to prompt your thinking here:  Do you need to add something about how the staking key is obtained?  For example, is there a check that the assets locked in the swap came from an address with the same staking key?  If any staking key can be specified does this open up the possibility for attacks by malicious web wallets and franken addresses somehow?

Author: When spending a swap UTxO, the staking credential is gotten directly from the UTxO itself. This information is passed to the script by the node itself and is unspoofable. So there is no risk from malicious users somehow providing an incorrect staking credential.

As far as where the assets go, I had originally required the assets go to the same address as the origin UTxO but, since Alice is explicitly approving the transaction, this seemed unnecessary. What if Alice was just moving her swap to a swap address with a different staking credential? Requiring the assets be locked at the same address as the origin means the swap would first have to be closed in one tx and then opened in the new address in another tx. I don't see any security issues with allowing the assets to go directly to another swap address. The requirement that the beacons must always go to a swap address is still properly enforced.

Reviewer 3: So it would be impossible to create a swap from an address without staking key (ie: address only)?

I don't see a problem with that though.

Author: > So it would be impossible to create a swap from an address without staking key (ie: address only)?

It is possible to create a UTxO at this address but it is not possible to tag it with the beacons. Addresses cannot *deny* UTxO creation just like stake pool operators cannot deny delegation. 

Since there is no staking credential, there is effectively no "admin" for the address. I did not want to have any locked assets so I set this edge case to always be spendable by anyone. This address is effectively first come, first serve. I don't think this address should ever be used though which is why the beacons cannot go to an address without a staking credential.

Once the beacons are minted to the swap address, the spending script does not allow closing the swap
*unless* the beacons are being burned. This prevents beacons from being sent to unrelated addresses
and guarantees the integrity of the off-chain queries.

All swaps must be stored with the proper beacons. Due to this requirement, the absolute minimum

> possible value for a swap UTxO to have is about 2 ADA. This minimum value can be reclaimed upon

Author: That seems strange to me. What about this sentence:
> All prices for ADA must be in units of Lovelace.

Should "ADA" and "Lovelace" be lowercase?

> possible value for a swap UTxO to have is about 2 ADA. This minimum value can be reclaimed upon

Reviewer: This ADA I didn't suggest lowercasing since the user is quoting an exchange price & therefore it is effectively a trading symbol & proper for it to be in full caps.  But you are right that lovelace is also lowercase... I didn't notice this because of having seen the proper name in initial caps so often in other contexts.

closing the swap so it should be thought of as a deposit. **All open swaps require a deposit of at
least 2 ADA.** While this behavior is due to the current protocol parameters, it is actually
a good thing since it helps prevents denial-of-service attacks from having a lot of "zombie swaps"
open (i.e., UTxOs where there is not enough of the offer asset for other users to swap).

It is possible to open swaps for multiple different trading pairs in a single transaction. The
minting policy is capable of still ensuring that all beacons are stored properly (with the proper
value and datum). For example, since each beacon's datum is specific to those beacons, mixing up
beacons and datums will cause the minting transaction to fail.

The `CreateSwap` allows also burning any beacons to enable composition with the `CloseOrUpdate`
redeemer. More on this later.

##### Closing or Updating a Swap

Open Swap UTxOs can be closed or updated by the address owner (signified by the address' staking
credential).

The `CloseOrUpdate` redeemer allows the owner to recover the deposit stored with the swap. **In

> order to reclaim the deposit, the beacons must be burned.** As the redeemer name suggests, swaps

Reviewer 3: Is it possible to have a eUTxO with only the 2 Ada deposit in it because the swap tokens have been completely spent already?  Does this mean that beacon tokens can remain until users reclaim their 2 Ada deposit?

Author: > Is it possible to have a eUTxO with only the 2 Ada deposit in it because the swap tokens have been completely spent already?

This wording isn't accurate since the UTxO should also have the assets that were asked for (eg, 2 ADA + 10 DJED after the USDC was taken). But it is possible for "finished" swaps to still have beacon tokens. These swaps are no longer swappable.

This is not an issue though because, when you query the beacons, you can filter out any UTxOs that do not have the offer asset. Koios allows doing this extra filtering server-side so the burden is not placed on the end user. The `cardano-swaps` CLI uses Koios like this. Blockfrost does not support server-side filtering which is why I did not add support for it in the CLI.

can also be updated inplace instead. The requirements for successfully using the `CloseOrUpdate`
redeemer are:

1) The beacons must go to an addres protected by the one-way swap validator script.
2) The beacons must go to an address using a valid staking credential.
3) The UTxOs with the beacons must have the proper value:
    - Exactly two kinds of beacons: pair beacon and offer beacon.
    - The beacons must correspond to the beacons in the datum.
    - There must be exactly 1 of each beacon.
    - No extraneous assets are in the UTxO. ADA is always allowed.
4) The beacons must be stored with the proper inline `SwapDatum`:
    - `beaconId` == this policy id.
    - `pairBeacon` == pair beacon asset name for this swap.
    - `offerId` == asset_id of the offer asset.
    - `offerName` == asset_name of the offer asset.
    - `offerBeacon` == offer beacon asset name for this swap's offer asset.
    - `askId` == asset_id of the ask asset.

>     - `askName` == asset_name of the ask asset.

Author: Those don't refer to actual variables so I'm not sure surrounding them with back-ticks is appropriate. I'll change them to "policy id" and "asset name" instead.

validator first checks for the trading pair beacon, each execution is dedicated to a specific
trading pair. Any other outputs are ignored in this specific execution. This logic works because a
script is executed once for every UTxO spent from the address. If input 1 is for beacon XYZ and
input 2 is for beacon ABC, the first execution can be dedicated to beacon XYZ and the second
execution can be dedicated to ABC. The net transaction will only succeed if all executions succeed.
This behavior allows cheaply composing swaps of different trading pairs that are located at the same
address. In other words, the design is taking advantage of the redundant executions.

At a low-level, for a swap execution to be successfull, all of the following must be true:

1. The input must contain the beacon for that trading pair - the required beacon is gotten from the
   datum for that UTxO. If the beacon is present, that means the datum can be trusted.
2. There must be an output to this address with the proper value and inline `SwapDatum`:
    - Must contain exactly the same value as the input except for the ask asset, offer asset, and

>     ADA.

Reviewer: yes, exchange rates or as a similar exchange symbol like currency quotes and values.  Generally it is capitalised because it's symbolic.  Here's our operative explanation on the Developer Portal (which unfortunately doesn't include Ada for the crypto itself: e.g. "Lebanon parliament voted to make Ada legal tender"): https://developers.cardano.org/docs/portal-style-guide/#editorial-style-guide

Anyone can use the `ForwardSwap` or `ReverseSwap` redeemers as long as the swap conditions are met.

##### Datums
Inline datums are used for all UTxOs. All UTxOs get the same type of datum:

``` Haskell
data SwapDatum = SwapDatum
  { beaconId :: CurrencySymbol -- ^ The beacon policy id for two-way swaps.
  , pairBeacon :: TokenName -- ^ The asset name for the beacon for this trading pair.
  , asset1Id :: CurrencySymbol -- ^ The policy id for the first asset in the sorted pair.
  , asset1Name :: TokenName -- ^ The asset name for the first asset in the sorted pair.
  , asset1Beacon :: TokenName --^ The asset name for the asset1 beacon.
  , asset2Id :: CurrencySymbol -- ^ The policy id for the second asset in the sorted pair.
  , asset2Name :: TokenName -- ^ The asset name for the second asset in the sorted pair.
  , asset2Beacon :: TokenName --^ The asset name for the asset2 beacon.
  , forwardPrice :: Rational -- ^ The swap price as a fraction: Asset1/Asset2.
  , reversePrice :: Rational -- ^ The swap price as a fraction: Asset2/Asset1.
  , prevInput :: Maybe TxOutRef -- ^ The output reference for the corresponding swap input.
  }
```

> The beacon information in the datum prevents misuse of beacons. The validator forces all beacons to

Author: I'll reference the one-way section. I think this makes adding new swaps easier in the case where the new swap type is very different than the current two types (an initial section would be inaccurate for the new type).

    - The UTxO must have asset1 and/or asset2.
    - No extraneous assets are in the UTxO. ADA is always allowed.
4) The beacons must be stored with the proper inline `SwapDatum`:
    - `beaconId` == this policy id.
    - `pairBeacon` == trading pair beacon asset name for this swap.
    - `asset1Id` == asset_id of asset1 for that trading pair.
    - `asset1Name` == asset_name of asset1 for that trading pair.
    - `asset1Beacon` == asset1 beacon asset name.
    - `asset2Id` == asset_id asset2 for that trading pair.
    - `asset2Name` == asset_name asset2 for that trading pair.
    - `asset2Beacon` == asset2 beacon asset name.
    - `forwardPrice` denominator > 0
    - `forwardPrice` > 0
    - `reversePrice` denominator > 0
    - `reversePrice` > 0
    - asset1 < asset2

The validator will assume that the trading pairs are sorted which is why asset1 must be less than
asset2. Asset1 and asset2 cannot be the same asset.

Once the beacons are minted to the swap address, the spending script does not allow closing the swap
*unless* the beacons are being burned. This prevents beacons from being sent to unrelated addresses
and guarantees the integrity of the off-chain queries.

> All swaps must be stored with the proper beacons. Due to this requirement, the absolute minimum

Author: > I believe it will improve the quality of this material as a reference

I'm disagree with this. The spec is not directed towards the laymen; it is directed towards developers (this is partly why I am okay with people skipping it). With that in mind, if I were going through the spec and trying to implement it myself, I would prefer to have everything I need in one location. Jumping around too much can make it difficult to keep track of where I am in the spec.

>  I believe it will improve ... its readability

I agree with this. Duplication can make a document feel dry. To me, there is a trade-off between being a good reference material and being easy to read. I don't think you can have both.

@zhekson1 and I have gone back and forth on a similar topic: since there is significant overlap in the architecture of the P2P protocols and therefore, significant overlap in the features, should those things be extracted out into a separate location like zhekson's defi-protocols [repo](https://github.com/zhekson1/CSL-DeFi-Protocols)? Personally, I am against extracting things out; I would rather duplicate things to have a single location with everything needed to understand it. I think part of the problem of showing people "Why Cardano?" is similar: there is no single location that has what you need. You need to jump around a lot to get the full picture.

> All swaps must be stored with the proper beacons. Due to this requirement, the absolute minimum

Author: I decided to keep the duplication in the spec for the reasons outlined above. If it does become to difficult to maintain, I can change it in the future.

3. Offered asset taken * price <= asked asset given.
4. Only the offered asset can leave and only the ask asset can be deposited. ADA can always be
deposited in case the minUTxOValue increased.

Requirement 1 guarantees that all invalid UTxOs (those missing beacons) belong to the address owner
and that swap inputs have a valid price: denominator > 0 and price > 0.

Requirements 2 & 4 guarantee that beacons from other trading pairs cannot be combined into one output
UTxO. This has two beneficial consequences:

1) The swap logic is very simple.
2) All swap UTxOs are as small as possible which maximizes how many swaps can be composed in a
single Tx as well as how easily the beacons can be queried.

Custom error messages are included to help troubleshoot why a swap failed.

## Benchmarks and Fee Estimations (YMMV)

The protocol is capable of handling 11-14 swaps in a single transaction, regardless of the
composition of one-way and two-way swaps in the transaction.

**No CIPs or hard-forks are needed. This protocol works on the Cardano blockchain, as is.**

Full benchmarking details can be found in the [Benchmarks](./Benchmarks/) folder.

## Features Discussion

> ### Full Delegation Control

Author: How's this section that would go before the delegation control section:

> ### Full Custody
> Since users get their own DEX addresses which use their own staking credentials for spending authorization, users maintain full custody of their assets at all times while using the DEX. *Not your keys, not your crypto.*

> ### Full Delegation Control

Reviewer: Excellent: exactly what I had in mind 🤩

> ### Full Delegation Control

Reviewer 3: I agree.  This feature is critical to the proper decentralised control of Cardano through the staking mechanism and voting in governance actions.  If users need to give up voting + staking control over their assets in order to use them in swaps or as collateral, then this severely impacts Cardano's resilience.

Sarah gives Charlie 10 HOSKY and receives 10 AGIX.
Sarah gives Bob 5 AGIX and receives 10 DUST.
Sarah gives Alice 10 DUST and receives 10 ADA.
Sarah now has her original 10 ADA plus an additional 5 AGIX.
This all occurs in one transaction where Sarah pays the transaction fee.
```

On net, Sarah pays the transaction fees and receives 5 AGIX in return, while four swaps are
fulfilled. As shown in this example, Sarah fulfills *both* the AGIX/DUST swap and the HOSKY/AGIX
swap by "passing through" those pairs on her way back to ADA. As long as the entry and exit pairs
(in this case ADA/HOSKY and DUST/ADA) have enough liquidity, arbitragers can spread that liquidity
into less liquid swap pairs. And since two-way swaps naturally incentivize providing liquidity for
major trading pairs, this requirement for entry and exit liquidity is naturally incentivized to be
satisfied.

Since the current design is capable of handling 11-14 swaps in a single transaction, there is ample
flexibility for finding arbitrage opportunities. For example, maybe the 3rd swap is at a slight loss
but the 5th swap is such a good arbitrage that it more than makes up for it resulting in a net

> profit for the arbitrage.

Reviewer 3: This concept of emergent liquidity is a massive opportunity and is enabled by the ease of composing a chain of dependent swaps.  As you indicate in your example, I think this will result in compression of spreads for actively traded pairs in exchange for execution of less liquid pairs.

people tend to to be loss averse ([prospect theory](https://en.wikipedia.org/wiki/Prospect_theory)).
A regulated institution (especially a publicly traded one) will need to always be able to manage any
risk of loss. Yield farming, where the institution is going to get paid a useless token, is simply
not going to cut it.

The fact that Cardano-Swaps prioritizes risk management over unbounded profits is a feature, not a
bug.

### If all users share a spending script, how are their assets protected?

The spending script gets the staking credential from the address of the UTxO being spent at
run-time. When an owner related action is being performed (closing or updating positions), the
spending script requires that the staking credential "signals approval" of the action:

- If the staking credential is a pubkey, the staking pubkey must sign the transaction.
- If the staking credential is a script, the script must be executed in the same transaction.

The staking credential effectively *becomes* the "owner" for all actions except for the actual swap
execution, in which case the spending credential is used directly.

> It is possible to execute a staking script even if 0 ADA is withdrawn from a reward address. The

Reviewer: ```suggestion
It is possible to execute a staking script even if 0 ada is withdrawn from a reward address. The
```

only requirement to use staking scripts like this is that the associated stake address must be
registered and delegated. Stake addresses can be utilized this way as soon as the
registration+delegation transaction is added to the chain. There is no epoch waiting-period.

### If Cardano-Swaps reaches mass adoption, won't TVL on Cardano go down?

Yes. Yes it will. TVL is a silly metric. It is a measure of who can be most inefficient with DeFi

> capital.

Reviewer: 😎😎😎👍👍👍

Reviewer 3: :100: :+1: :+1: :+1:
